Fix GHCJS for cabal-install-1.22.8.0

cstrahan opened the pull request:
This splits off the stage2 packages from `ghcjs`. As a result, this _should_ fix the build (currently rebuilding this on the latest master).

/cc @Ericson2314 @pikajude @ryantrinkle

cstrahan committed:
ghcjs: fix building with cabal-install-1.22.8.0

cstrahan committed:
ghcjs: use native ghc to build Setup.hs

Ericson2314 commented:
Just wanted to say that I've discussed it with @cstrahan and it would be good for this to be merged first, and me to rebase my more radical https://github.com/NixOS/nixpkgs/pull/13074 on top of it.

cstrahan committed:
ghcjs: don't propagate Cabal to all packages

Previously, we were compiling Setup.hs with ghcjs (instead of  ghc),
and that required that Cabal be available, otherwise the Setup.hs would
fail to compile.

Now that we are compiling Setup.hs with ghc, it's no longer necessary
to add Cabal as a dependency to all ghcjs packages.

peti commented:
The changes look good to me. :+1:

cstrahan committed:
ghcjs: replace integer-simple with integer-gmp

GHCJS uses integer-gmp, but cabal2nix generates a dependency list that
includes integer-simple instead. This tweaks the stage2 generator to
replace any instance of integer-simple with integer-gmp.

Things currently still work without this change (assuming
integer-simple is defined as null), as ghcjs includes integer-gmp in
its stage1 packages - so it's always available.

However, this change makes things a bit more explicit, rather than
leaving things to chance. If at any point the stage1 packages are also
split up into separate derivations, the integer-gmp package will need to
be passed along to the packages that depend on it. This change should
prevent some confusion going forward.

cstrahan committed:
ghcjs-shims: 2016/01/07 -> 2016/02/10

Bump revisions: 45f44f5f027 -> 4df1808d031

cstrahan commented:
Thanks for reviewing!
